elastic/kibana
# [Upgade Assistant] Enable UA for minor upgrades

# Summary

UA is always enabled, the customer needs to check if their cluster needs an upgrade before they start addressing warnings and errors. This way they can also proactively address errors and warnings preventing upgrades before a new stack release is even available.

Currently, we are not pinging an external API to check if the cluster is not at the latest stack version hence this step must be manually done by the users.

- [x] Enable UA for minor upgrades
- [x] New UA design changes
- [x] Update docs links
- [x] Connect ES upgrade readiness apis
- [x] Review new UA interface copy
- [x] Unit and functional tests

## Details and screenshots:

### Updated UA text to adapt to any release upgrade not just `8.latest`
The documentation links will redirect the user to the `current` version of the stack instead of `8.0`. This way the user will check if they are not on the latest version and follow through with the upgrade steps.

<img width="949" alt="image" src="https://user-images.githubusercontent.com/6191849/196455440-c464b95a-232d-4963-8eb1-047f7f77db09.png">

<img width="1266" alt="image" src="https://user-images.githubusercontent.com/6191849/199257940-2319bec6-f903-4ecc-b0ab-c1e0e6144fc3.png">


### Upgrade readiness checks
The ES upgrade readiness checks are integrated into their deprecations API.
Initially we had designs to keep this in a separate step, however merging the checks into the already existing third step provides a better UX and a shorter barrier to upgrade

New checks:
- Critical deprecation warning for potentially exceeding maximum open shards setting https://github.com/elastic/elasticsearch/issues/85702 
- Critical deprecation warning for low disk watermark https://github.com/elastic/elasticsearch/pull/83544

### Migrating system indices
Migrating system indices are only required during major upgrades. This step will be marked as complete to the users unless they are upgrading to a major version. On non-major version upgrades, we will hide this step completely.

# Manual testing

Since the ES changes are using the same deprecation API there is no need to add tests from our side to check that they are getting triggered correctly outside a couple of surface-level checks.

## Minor version upgrade

1. Launch Kibana
2. Go to advanced settings / upgrade assistant
3. Check upgrade steps

## Upgrade Readiness
The ES Upgrade readiness checks are integrated into the ES deprecations API and will show up on the Elasticsearch deprecations section inside the Upgrade Assistance.

## Low disk watermark

1. Reduce the low watermark disk threshold to mimic disk usage exceeding low watermark. Note that for ES running in dev mode the disk threshold is disabled hence we'd need to enable it as well.

```
PUT _cluster/settings
{
  "transient": {
    "cluster.routing.allocation.disk.watermark.low" : "30%",
    "cluster.routing.allocation.disk.threshold_enabled": true,
    "cluster.info.update.interval": "10s"
  }
}
```

2. Call the ES deprecations API
```
GET /_migration/deprecations
```

3. Check the result
```
{
  "cluster_settings": [],
  "node_settings": [
    {
      "level": "critical",
      "message": "Disk usage exceeds low watermark",
      "url": "https://ela.st/es-deprecation-7-disk-watermark-exceeded",
      "details": "Disk usage exceeds low watermark, which will prevent reindexing indices during upgrade. Get disk usage on all nodes below the value specified in cluster.routing.allocation.disk.watermark.low (nodes impacted: [Bamieh-MBP])",
      "resolve_during_rolling_upgrade": false
    }
  ],
  "index_settings": {},
  "ml_settings": []
}
```

4. Open the upgrade assistant to check the deprecation

<img width="664" alt="image" src="https://user-images.githubusercontent.com/6191849/199257074-e3073413-6825-4e57-b37b-29a8a5382003.png">

<img width="1725" alt="image" src="https://user-images.githubusercontent.com/6191849/199257025-53223102-a8d6-4934-a6d9-7f42e8df4e9c.png">

## Total shards check

1. Reduce the max shards per node to mimic the deprecation when the cluster has too many shards
```
PUT /_cluster/settings
{
  "persistent" : {
    "cluster.max_shards_per_node" : "9"
  }
}
```

2. Call the ES deprecations API
```
GET /_migration/deprecations
```

3. Check the result
```
{
  "cluster_settings": [
    {
      "level": "warning",
      "message": "The cluster has too many shards to be able to upgrade",
      "url": "https://ela.st/es-deprecation-8-shard-limit",
      "details": "Upgrading requires adding a small number of new shards. There is not enough room for 10 more shards. Increase the cluster.max_shards_per_node setting, or remove indices to clear up resources.",
      "resolve_during_rolling_upgrade": false
    }
  ],
  "node_settings": [],
  "index_settings": {},
  "ml_settings": []
}
```

4. Open the upgrade assistant to check the deprecation
<img width="662" alt="image" src="https://user-images.githubusercontent.com/6191849/199254045-c49a52f5-6021-4020-9131-51c72ed853e5.png">

<img width="1728" alt="image" src="https://user-images.githubusercontent.com/6191849/199259679-2f06287b-6bd0-4f15-814e-4d509deda0f2.png">
